### PR TITLE
Update .NET version checks to be more correct when Windows services when installing master and agent on platforms with modern .NET versions

### PR DESCRIPTION
**Pull request by nxtn**
* We do not require JIRA issues for minor improvements.

No functional changes will be involved if this class is only used internally.

- .NET 1.0 and .NET 1.1 are compatible.
- .NET 2.0, .NET 3.0 and .NET 3.5 are compatible.
- .NET 4.0 and later versions are compatible.

For more information on version checks, see [How to: Determine which .NET Framework versions are installed](https://docs.microsoft.com/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed).

This change has no tests because CI runs on Linux.

### Proposed changelog entries

* Update supported .NET version checks to be more correct when Windows services when installing master and agent on platforms with modern .NET versions

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

/cc @oleg-nenashev 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
CC @jenkinsci/sig-platform

**Comment by NextTurn** [comment]
I didn't see any test for these JNA classes. Do you want some tests for this?

**Comment by NextTurn** [comment]
External uses of this class to be updated first:

https://github.com/jenkinsci/windows-slaves-plugin/blob/0e4a473dc231ea69b8b57d0a9707d32a39c8cfab/src/main/java/hudson/os/windows/ManagedWindowsServiceLauncher.java#L286

https://github.com/jenkinsci/windows-slave-installer-module/blob/bc6042c468609a70108c525ba4197cf878243b72/src/main/java/org/jenkinsci/modules/windows_slave_installer/WindowsSlaveInstaller.java#L114

**Review by oleg-nenashev — APPROVED** [review]
This is a case when maybe just dropping the check might be preferable. Windows Service Wrapper supports all modern platforms at the moment. OTOH I approve it, because the check might be important again if we ever decide to drop .NET 2.0 and .NET 3.5 support in WinSW (FTR https://github.com/kohsuke/winsw/issues/434 )

There is also an option to move the entire check logic to WinSW directly, because it might be helpful to other users. E.g. we could invoke `winsw.exe check-platform` which would verify the system compatibility with the wrapper. The downside is that winsw would have to be runnable on lower .NET versions, and it is generally not possible unless you want to write unsafe main code or accept cryptic .NET startup messages (@NextTurn please correct me if I am wrong). A separate `winsw-checker.exe` binary could be an option.

**Comment by NextTurn** [comment]
> There is also an option to move the entire check logic to WinSW directly, because it might be helpful to other users. E.g. we could invoke `winsw.exe check-platform` which would verify the system compatibility with the wrapper. The downside is that winsw would have to be runnable on lower .NET versions, and it is generally not possible unless you want to write unsafe main code or accept cryptic .NET startup messages (@NextTurn please correct me if I am wrong). A separate `winsw-checker.exe` binary could be an option.

That's such an interesting idea. But a much easier way is to run `winsw.exe version` and see if it crashes. ;)

**Comment by oleg-nenashev** [comment]
> That's such an interesting idea. But a much easier way is to run winsw.exe version and see if it crashes. ;)

"... or accept cryptic .NET startup messages" in my previous comments :) . Recent .NET versions generally handle such situations better than Java 1.8 on default settings, but the error messages are likely to be unfriendly to users

**Comment by NextTurn** [comment]
> The downside is that winsw would have to be runnable on lower .NET versions, and it is generally not possible unless you want to write unsafe main code

Technically it's impossible. This is similar to .NET Core runtime-dependent executable scheme, where a native executable finds and hosts CLR with managed assemblies bundled as PE resources.

I believe it tasks less time to rewrite the whole program in C/C++. :)

File: core/src/main/java/hudson/util/jna/DotNet.java [hunk 1]
```diff
@@ -30,83 +30,232 @@
...
 
-    private static boolean matches(String keyName, int major, int minor) {
-        Matcher m = VERSION_PATTERN.matcher(keyName);
-        if(m.matches()) {
-            int mj = Integer.parseInt(m.group(1));
-            if(mj>=major) {
-                int mn = Integer.parseInt(m.group(2));
-                if(mn>=minor)
-                    return true;
+    private static boolean isV45PlusInstalled(int minor, IJIWinReg registry, JIPolicyHandle hklm) throws JIException {
```
[the remark this answers is not in the document]
Comment (NextTurn) [reply core/src/main/java/hudson/util/jna/DotNet.java]: Updated. I've never seen a library any uglier than this *org.jinterop.winreg*.

**Comment by oleg-nenashev** [comment]
So, would it make sense to just drop the check for now to keep the code simple?

**Comment by NextTurn** [comment]
> So, would it make sense to just drop the check for now to keep the code simple?

The code isn't complicated. It may be helpful to implement install time telemetry.

**Comment by oleg-nenashev** [comment]
> The code isn't complicated. It may be helpful to implement install time telemetry.

My bet is that we would discover that almost nobody uses this feature. There is an MSI installer which does Windows service installation from outside.

In order to use the Windows installer as is, one would need to run the jenkins.war file as Administrator, and it it not what I would recommend in good conscience. Without Administrator Jenkins will just generate the WinSW configuration files and fail, and then the user will have to run `jenkins.exe` manually. Once WinSW supports uiAccess, the behavior will be more user-friendly (FTR https://github.com/kohsuke/winsw/pull/410/files#r379879270).

I am fine with merging the code as is anyway

**Comment by NextTurn** [comment]
FYI 

> If an application presents a UIAccess attribute when it requests privileges, the application is stating a requirement to bypass UIPI restrictions for sending messages across privilege levels. Devices implement the following policy 
checks before starting an application with UIAccess privilege.
> 
> 1.  The application must have a digital signature that can be verified by using a digital certificate that is associated with the Trusted Root Certification Authorities store on the local computer.
> 2.  The application must be installed in a local folder that is writeable only by administrators, such as the Program Files directory. The allowed directories for UI automation applications are:
> 
>     1.  %ProgramFiles% and its subdirectories.
>     2.  %WinDir% and its subdirectories, except a few subdirectories that are excluded because standard users have write access.

**Review by slide — APPROVED** [review]

File: core/src/main/java/hudson/util/jna/DotNet.java [hunk 2]
```diff
@@ -30,83 +30,232 @@
...
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
 
 /**
  * .NET related code.
  * 
  * @author Kohsuke Kawaguchi
  */
 public class DotNet {
+    private static final String PATH10 = "SOFTWARE\\Microsoft\\.NETFramework\\Policy\\v1.0\\3705";
```
[the remark this answers is not in the document]
Comment (NextTurn) [reply core/src/main/java/hudson/util/jna/DotNet.java]: Done for completeness. We could directly return `false` as well.

**Comment by NextTurn** [comment]
> Once WinSW supports uiAccess, the behavior will be more user-friendly (FTR https://github.com/kohsuke/winsw/pull/410/files#r379879270).

FYI 

> If an application presents a UIAccess attribute when it requests privileges, the application is stating a requirement to bypass UIPI restrictions for sending messages across privilege levels. Devices implement the following policy checks before starting an application with UIAccess privilege.
> 
> 1.  The application must have a digital signature that can be verified by using a digital certificate that is associated with the Trusted Root Certification Authorities store on the local computer.
> 2.  The application must be installed in a local folder that is writeable only by administrators, such as the Program Files directory. The allowed directories for UI automation applications are:
> 
>     1.  %ProgramFiles% and its subdirectories.
>     2.  %WinDir% and its subdirectories, except a few subdirectories that are excluded because standard users have write access.

It's unlikely that we request the users to place the files in secure locations if they choose not to use the MSI installer.

**Comment by NextTurn** [comment]
FTR here are all the UIAccess applications in secure locations on my machine.

```
C:\Program Files\Common Files\microsoft shared\ink\TabTip.exe
C:\Program Files (x86)\Common Files\Microsoft Shared\ink\TabTip32.exe
C:\Program Files (x86)\Common Files\Microsoft Shared\VS7Debug\VsImmersiveActivateHelper.exe
C:\Windows\System32\ctfmon.exe
C:\Windows\System32\EaseOfAccessDialog.exe
C:\Windows\System32\Magnify.exe
C:\Windows\System32\msra.exe
C:\Windows\System32\Narrator.exe
C:\Windows\System32\osk.exe
C:\Windows\System32\psr.exe
C:\Windows\System32\quickassist.exe
C:\Windows\System32\rdpinput.exe
C:\Windows\System32\RdpSa.exe
C:\Windows\System32\rdpshell.exe
C:\Windows\System32\oobe\AuditShD.exe
```

**Review by oleg-nenashev — APPROVED** [review]
I suggest to move it forward. The current behavior is net improvement being compared to the previous one, we can always revise it later it if needed.  I plan to merge it in 24 hours if no negative feedback

**Review by timja — APPROVED** [review]

**Pull request merged**